Automattic/jetpack
# CRM: Cleaning up WP Editor helper functions and wp_editor usage

## Proposed changes:

* All `zeroBSCRM_io_WPEditor` helper functions have been removed in this PR: `zeroBSCRM_io_WPEditor_WPEditorToDB`, `zeroBSCRM_io_WPEditor_DBToWPEditor`, `zeroBSCRM_io_WPEditor_DBToHTML`, `zeroBSCRM_io_WPEditor_DBToHTMLExcerpt`. As such, any place where they were called has been cleaned up to work without them, with replacement functions added where needed.
  * The `DBToWPEditor`, `DBToHTML`, and `DBToHTMLExcerpt` helper functions were all decoding HTML, which is not the expected behaviour in TinyMCE editors. The areas where these functions were being used have now been changed to reflect what is expected in those cases.
* Similarly, `wp_editor` function usage has also been cleaned up. In two cases we previously overcomplicated the `wp_editor` usage - in [this example](https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/crm/admin/email/main.page.php?rgh-link-date=2023-04-25T15%3A27%3A42Z#L303) plus [this example](https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/crm/admin/email/main.page.php?rgh-link-date=2023-04-25T15%3A27%3A42Z#L558), `$content` is always an empty string (it's the default info in the editor.) That has been changed as well. In the remaining case where a change was made, decoding wasn't needed.


### Other information:


## Jetpack product discussion

https://github.com/Automattic/zero-bs-crm/issues/2662

## Does this pull request change what data or activity we track or use?

No.

## Testing instructions:

To test the `wp_editor` changes:

* The two `wp_editor` changes affecting the email page affect content added to the editor before you start typing - and in both cases this is set to nothing (so it doesn't need decoding). 
* To confirm the editor isn't affected by the changes, on the email page - `/wp-admin/admin.php?page=zerobscrm-emails` - click on compose and manually add some HTML to the editor. Alongside that, use the editor mark up some text (eg bold, italic).
* Click send and check the content of the email in the sent folder. The HTML that was manually added should show as HTML (unrendered).
* Do the same but from within the thread of a sent email, the results should be the same.
* This behaviour should be the same both before and after this PR is applied. 
* The third `wp_editor` change affects [email templates](https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/crm/includes/ZeroBSCRM.AdminPages.php#L1192)
* To test this, visit the email templates page: `wp-admin/admin.php?page=zbs-email-templates&zbs_template_id=1`. 
* Add some acceptable HTML (another bold title for example), and preview the email. The HTML should render as before. 
* Add some unacceptable HTML (a script tag), that will be stripped out. Both before and after previewing this PR, the results should be the same.
* Open a quote proposal in the Client Portal - and the result should be consistent with previous changes (manual HTML not rendered).

To test the removal of the ZBS helper functions:
* On the email page, open a sent email where some manual HTML has been added to an email body. The HTML should now not be rendered (it was before). Using the editor toolbar itself to add the HTML should still render the HTML.
* Check email excerpts - no HTML should be rendered.
* With Quote templates - create a new template and manually add HTML in the visual editor. When adding that quote template to a quote, the HTML should not render. If you add the HTML using the editor, it should render as expected. Placeholders shouldn't be affected.
* Similarly, add manual HTML to the quote editor (visual editor) and save it - it shouldn't render unless using the toolbar to add HTML.
